File: stdlib/js/hull/web/cookie.js
/**
 * @file hull:web:cookie
 * @module hull:web:cookie
 * @description HTTP cookie parsing and serialization.
 *
 * Lua parity: same surface as `hull.web.cookie` with camelCase options.
 *
 * @license AGPL-3.0-or-later
 */

/**
 * Parse a `Cookie` header string into a name-value object.
 *
 * Empty or non-string input returns an empty object (never throws).
 *
 * @param {string|null|undefined} headerString  Value of the inbound `Cookie` header.
 * @returns {Object.<string, string>}  Parsed cookies. `{}` when input is empty.
 *
 * @example
 * const cookies = cookie.parse(req.headers.cookie);
 * if (cookies.session_id) { ... }
 */
function parse(headerString) {
    const result = {};
    if (!headerString || typeof headerString !== "string")
        return result;

    const pairs = headerString.split(";");
    for (let i = 0; i < pairs.length; i++) {
        const pair = pairs[i].trim();
        if (pair.length === 0) continue;

        const eqIdx = pair.indexOf("=");
        if (eqIdx < 0) continue;

        const name = pair.substring(0, eqIdx).trim();
        if (name.length === 0) continue;

        let value = pair.substring(eqIdx + 1).trim();

        // Strip surrounding quotes if present
        if (value.length >= 2 && value[0] === '"' && value[value.length - 1] === '"')
            value = value.substring(1, value.length - 1);

        // Decode percent-encoded values
        try {
            result[name] = decodeURIComponent(value);
        } catch (e) {
            result[name] = value;
        }
    }

    return result;
}

/**
 * Serialize a cookie into a `Set-Cookie` header value.
 *
 * @param {string} name           Cookie name.
 * @param {string} value          Cookie value.
 * @param {Object} [opts]         Options.
 * @param {string}  [opts.path="/"]
 * @param {boolean} [opts.httpOnly=true]
 * @param {boolean} [opts.secure=true]
 * @param {"Lax"|"Strict"|"None"} [opts.sameSite="Lax"]
 * @param {number}  [opts.maxAge]
 * @param {string}  [opts.domain]
 * @param {string}  [opts.expires]  RFC 1123 date.
 * @returns {string}  `Set-Cookie` header value. Pair with `res.header("Set-Cookie", v)`.
 *
 * @example
 * res.header("Set-Cookie", cookie.serialize("session", sid, { sameSite: "Strict" }));
 */
function serialize(name, value, opts) {
    if (!name || typeof name !== "string")
        throw new Error("cookie name is required");
    if (/[=;,\s\x00-\x1f\x7f]/.test(name))
        throw new Error("cookie name contains invalid characters");
    if (value !== null && value !== undefined && /[\x00-\x1f\x7f;]/.test(String(value)))
        throw new Error("cookie value contains invalid characters");

    const o = opts || {};

    // Percent-encode the value
    let encoded;
    if (value === null || value === undefined)
        encoded = "";
    else
        encoded = encodeURIComponent(String(value));

    let str = name + "=" + encoded;

    // Path (default: "/"). Reject control bytes and `;` so a
    // user-supplied path can't inject directives (e.g.
    // `; HttpOnly=false`) or split the Set-Cookie header.
    const path = o.path !== undefined ? o.path : "/";
    if (path) {
        if (/[;\r\n\x00]/.test(path))
            throw new Error("cookie: invalid path (contains ';', "
                + "CR, LF, or NUL)");
        str += "; Path=" + path;
    }

    // Domain
    if (o.domain) {
        if (!/^[a-zA-Z0-9._-]+$/.test(o.domain))
            throw new Error("cookie: invalid domain");
        str += "; Domain=" + o.domain;
    }

    // MaxAge
    if (o.maxAge !== undefined && o.maxAge !== null) {
        const maxAge = Math.floor(o.maxAge);
        if (isNaN(maxAge))
            throw new Error("maxAge must be a number");
        str += "; Max-Age=" + maxAge;
    }

    // Expires
    if (o.expires) {
        if (typeof o.expires === "string") {
            if (/[\x00-\x1f;]/.test(o.expires))
                throw new Error("cookie: invalid expires value");
            str += "; Expires=" + o.expires;
        } else if (typeof o.expires === "number")
            str += "; Expires=" + new Date(o.expires * 1000).toUTCString();
    }

    // HttpOnly (default: true)
    const httpOnly = o.httpOnly !== undefined ? o.httpOnly : true;
    if (httpOnly)
        str += "; HttpOnly";

    // Secure=true by default. Set secure=false explicitly for local HTTP dev.
    const secure = o.secure !== undefined ? o.secure : true;
    if (secure)
        str += "; Secure";

    // SameSite (default: "Lax")
    const sameSite = o.sameSite !== undefined ? o.sameSite : "Lax";
    if (sameSite) {
        const ss = String(sameSite);
        if (ss === "Strict" || ss === "Lax" || ss === "None")
            str += "; SameSite=" + ss;
        else
            throw new Error("sameSite must be Strict, Lax, or None");
    }

    return str;
}

// Whitelisted Set-Cookie attributes. Avoiding Object.assign here prevents
// prototype-pollution if the caller's opts object inherits adversarial keys.
const COOKIE_ATTRS = ["path", "domain", "httpOnly", "secure", "sameSite",
                       "maxAge", "expires"];

/**
 * Build a `Set-Cookie` value that deletes the named cookie.
 *
 * @param {string} name  Cookie name to clear.
 * @param {Object} [opts]  Same options as {@link serialize}; `path`/`domain` should match the original.
 * @returns {string}  `Set-Cookie` header value with `Max-Age=0`.
 */
function clear(name, opts) {
    const o = Object.create(null);
    if (opts) {
        for (const k of COOKIE_ATTRS) {
            if (Object.prototype.hasOwnProperty.call(opts, k))
                o[k] = opts[k];
        }
    }
    o.maxAge = 0;
    o.expires = "Thu, 01 Jan 1970 00:00:00 GMT";
    return serialize(name, "", o);
}

const cookie = { parse, serialize, clear };
export { cookie };

File: stdlib/lua/hull/web/cookie.lua
--- HTTP cookie parsing and serialization.
--
-- @module hull.web.cookie
-- @license AGPL-3.0-or-later

local cookie = {}

--- Parse a `Cookie` header string into a name-value table.
--
-- Each cookie is split on `=`, trimmed of surrounding whitespace.
-- Empty input or `nil` returns an empty table (never errors).
--
-- @tparam string|nil header_string  Value of the inbound `Cookie` header.
-- @treturn {[string]=string} Parsed cookies. Empty table when input is `nil` or empty.
-- @usage
-- local cookies = cookie.parse(req.headers["cookie"])
-- if cookies.session_id then ... end
function cookie.parse(header_string)
    local result = {}
    if not header_string or header_string == "" then
        return result
    end

    for pair in string.gmatch(header_string, "[^;]+") do
        -- Trim leading/trailing whitespace
        pair = pair:match("^%s*(.-)%s*$")
        if pair ~= "" then
            local eq = pair:find("=", 1, true)
            if eq then
                local name = pair:sub(1, eq - 1):match("^%s*(.-)%s*$")
                local value = pair:sub(eq + 1):match("^%s*(.-)%s*$")
                if name ~= "" then
                    result[name] = value
                end
            end
        end
    end

    return result
end

--- Serialize a cookie into a `Set-Cookie` header value.
--
-- @tparam string name   Cookie name.
-- @tparam string value  Cookie value. Caller is responsible for any encoding.
-- @tparam[opt] table opts  Options:
--
--   - `path`     (string, default `"/"`)
--   - `httponly` (boolean, default `true`)
--   - `secure`   (boolean, default `true`; set `false` explicitly for local
--                 HTTP dev). Secure-by-default matches the code.
--   - `samesite` (string, default `"Lax"`; one of `"Lax"|"Strict"|"None"`)
--   - `max_age`  (integer, seconds; omits attribute if `nil`)
--   - `domain`   (string)
--   - `expires`  (string, RFC 1123 date; omits if `nil`)
--
-- @treturn string  `Set-Cookie` header value. Pair with `res:header("Set-Cookie", v)`.
-- @usage
-- res:header("Set-Cookie", cookie.serialize("session", sid, { secure = true }))
function cookie.serialize(name, value, opts)
    opts = opts or {}

    -- Validate cookie name (RFC 6265 token)
    if not name or name == "" or name:find("[%c;= ,]") then
        error("cookie: invalid cookie name")
    end
    -- Reject control characters and semicolons in value
    if value and value:find("[%c;]") then
        error("cookie: invalid cookie value")
    end

    local parts = { name .. "=" .. (value or "") }

    -- Path (default "/")
    local path = opts.path
    if path == nil then path = "/" end
    if path then
        parts[#parts + 1] = "Path=" .. path
    end

    -- HttpOnly (default true)
    local httponly = opts.httponly
    if httponly == nil then httponly = true end
    if httponly then
        parts[#parts + 1] = "HttpOnly"
    end

    -- Secure=true by default. Set secure=false explicitly for local HTTP dev.
    local secure = opts.secure
    if secure == nil then secure = true end
    if secure then
        parts[#parts + 1] = "Secure"
    end

    -- SameSite (default "Lax")
    local samesite = opts.samesite
    if samesite == nil then samesite = "Lax" end
    if samesite then
        parts[#parts + 1] = "SameSite=" .. samesite
    end

    -- Max-Age
    if opts.max_age then
        parts[#parts + 1] = "Max-Age=" .. tostring(opts.max_age)
    end

    -- Domain - RFC 6265 attribute syntax. Accepts an optional leading "."
    -- (legacy subdomain-matching form, still in widespread use). Rejects
    -- length > 253, double dots, trailing dots, and any non-host characters.
    if opts.domain then
        local d = opts.domain
        if #d > 253
           or not d:match("^%.?[a-zA-Z0-9][a-zA-Z0-9%-%.]*[a-zA-Z0-9]$")
           or d:find("%.%.", 1, false) then
            error("cookie: invalid domain")
        end
        parts[#parts + 1] = "Domain=" .. d
    end

    -- Expires
    if opts.expires then
        if type(opts.expires) ~= "string" or opts.expires:find("[%c;]") then
            error("cookie: invalid expires value")
        end
        parts[#parts + 1] = "Expires=" .. opts.expires
    end

    return table.concat(parts, "; ")
end

--- Build a `Set-Cookie` value that deletes the named cookie.
--
-- Sets `Max-Age=0` and an empty value. The browser deletes the cookie
-- on receipt.
--
-- @tparam string name  Cookie name to clear.
-- @tparam[opt] table opts  Same options as @{cookie.serialize}; `path` /
--   `domain` should match the original to ensure the right cookie is cleared.
-- @treturn string  `Set-Cookie` header value.
-- @usage
-- res:header("Set-Cookie", cookie.clear("session", { path = "/" }))
function cookie.clear(name, opts)
    opts = opts or {}
    local clear_opts = {}
    -- Copy relevant options
    clear_opts.path = opts.path
    clear_opts.domain = opts.domain
    clear_opts.httponly = opts.httponly
    clear_opts.secure = opts.secure
    clear_opts.samesite = opts.samesite
    clear_opts.max_age = 0
    return cookie.serialize(name, "", clear_opts)
end

return cookie
